### PR TITLE
Stylesheet parsing logic improvements

### PR DESCRIPTION
**Pull request by bsweeney**
- Updates regex patterns to improve stylesheet ingestion
- Improves media query handling
  - Supports media queries with more than one condition
  - Attaches all media queries to a single Style object instead of replicating the Style object for each query
  - Expands supports logical operators (not, or)
  - Fills out the media query logic for at-import rules
- Improves url() CSS function handling
- Improves URI ("src" property) handling

### CODE REVIEW

**Review by Mellthas — CHANGES_REQUESTED** [review]

File: src/Css/Stylesheet.php [hunk 1]
```diff
@@ -1209,81 +1236,96 @@ private function _parse_css($str)
...
-            if ($match[2] !== "") {
+            if ($match["CSS_ATRULE_IDENTIFIER"] !== "") {
+                $atrule_identifier = strtolower($match["CSS_ATRULE_IDENTIFIER"]);
                 // Handle @rules
-                switch ($match[2]) {
+                switch ($atrule_identifier) {
 
                     case "import":
-                        $this->_parse_import($match[3]);
+                        $this->_parse_import($match["CSS_ATIMPORT_RULE"]);
```
[the remark this answers is not in the document]
Comment (bsweeney) [reply src/Css/Stylesheet.php:1290]: I think you're right.

File: src/Css/Stylesheet.php [hunk 2]
```diff
@@ -1301,33 +1293,35 @@ private function _parse_css($str)
...
-                                    if (empty($media_query_match["CSS_MEDIA_QUERY_TYPE"]) === false) {
-                                        $media_query_feature = strtolower($media_query_match["CSS_MEDIA_QUERY_TYPE"]);
-                                        $media_query_value = strtolower($media_query_match["CSS_MEDIA_QUERY_OP"]);
-                                        $mq[] = [$media_query_feature, $media_query_value];
-                                    } elseif (empty($media_query_match["CSS_MEDIA_QUERY_FEATURE"]) === false) {
-                                        $media_query_feature = strtolower($media_query_match["CSS_MEDIA_QUERY_FEATURE"]);
-                                        $media_query_value = (array_key_exists("CSS_MEDIA_QUERY_CONDITION", $media_query_match) ? strtolower($media_query_match["CSS_MEDIA_QUERY_CONDITION"]) : null);
-                                        $mq[] = [$media_query_feature, $media_query_value];
-                                    }
+                            $mq = [];
```
[the remark this answers is not in the document]
Comment (bsweeney) [reply src/Css/Stylesheet.php:1301]: Good catch. I must not have tested with a sufficiently complex media query.

File: src/Css/Stylesheet.php [hunk 3]
```diff
@@ -1449,43 +1443,107 @@ private function _parse_import($url)
...
+                        $media_query_operator = "not";
+                    }
 
-            // if URL is a CSS string, wrap it in the url function for parsing by the resolve_url method
-            if (mb_strpos($url, "url(") === false) {
-                $url = "url($url)";
-            }
+                    switch ($media_query_feature) {
+                        case "height":
+                            $feature_match = $paper_height === (float)$style->length_in_pt($media_query_value);
```
[the remark this answers is not in the document]
Comment (bsweeney) [reply src/Css/Stylesheet.php:1516]: How disappointingly sloppy of me.

**Comment by bsweeney** [comment]
Thanks for reviewing this. I definitely need a co-pilot next to me 😅

**Review by Mellthas — APPROVED** [review]
Looks like a nice set of improvements!

**Pull request merged**